saleor/warehouse/models.py

```diff
@@ -128,6 +137,13 @@ class Allocation(models.Model):
     )
     quantity_allocated = models.PositiveIntegerField(default=0)
 
+    def clean(self):
+        if self.order_line is None and self.checkout_line is None:
+            raise ValidationError("Order line or checkout line is required")
+
     class Meta:
-        unique_together = [["order_line", "stock"]]
```

Comment:
moving to UniqueConstraint as it's faster then unique_together, and also might be decaprated in the future

https://docs.djangoproject.com/en/3.0/ref/models/options/#django.db.models.Options.unique_together